libs/ballot-interpreter/src/hmpb-rust/timing_marks.rs

```diff
@@ -461,85 +462,106 @@ pub fn find_timing_mark_grid(
 
     for y in search_rect.top()..search_rect.bottom() {
         for x in search_rect.left()..search_rect.right() {
-            let luma = image.get_pixel(x as u32, y as u32);
-            if rect.contains(Point::new(x, y)) {
-                if luma.0[0] <= threshold {
-                    black_pixel_count += 1;
+            let point = Point::new(x, y);
+            if image_rect.contains(point) {
+                let luma = image.get_pixel(x as u32, y as u32);
+                let expects_mark_pixel = expected_timing_mark_rect.contains(point);
+                let is_black_pixel = luma.0[0] <= threshold;
+
+                if expects_mark_pixel == is_black_pixel {
+                    if expects_mark_pixel {
+                        mark_pixel_match_count += 1;
+                    } else {
+                        padding_pixel_match_count += 1;
+                    }
                 }
-            } else if luma.0[0] > threshold {
-                white_pixel_count += 1;
             }
         }
     }
 
-    let black_pixel_score = black_pixel_count as f32 / black_pixel_search_area as f32;
-    let white_pixel_score = white_pixel_count as f32 / white_pixel_search_area as f32;
-
-    Some(UnitIntervalScore(black_pixel_score * white_pixel_score))
+    // Note that we DO NOT use the actual width and height of the timing mark
+    // here, but the expected width and height. This is because the timing mark
+    // may be cropped and its score will be artificially inflated if we use the
+    // actual width and height.
+    let timing_mark_area = expected_width * expected_height;
+    TimingMarkScore {
+        mark_score: UnitIntervalScore(mark_pixel_match_count as f32 / timing_mark_area as f32),
+        padding_score: UnitIntervalScore(
+            padding_pixel_match_count as f32 / (search_area - timing_mark_area) as f32,
+        ),
+    }
 }
 
-/// Filters the bottom timing marks by examining the image to determine if the
-/// timing marks are actually present.
+/// Filters the bottom timing marks by using the candidate timing mark scores to
+/// determine whether timing marks are actually present.
 pub fn find_actual_bottom_marks(
     complete_timing_marks: &Complete,
-    ballot_image: &BallotImage,
-) -> Vec<Option<Rect>> {
-    const MIN_REQUIRED_MARK_SCORE: UnitIntervalScore = UnitIntervalScore(0.5);
+) -> Vec<Option<CandidateTimingMark>> {
+    // allow marks at the exterior (i.e. corners) to be substantially cropped
+    const MIN_REQUIRED_EXTERIOR_MARK_SCORE: UnitIntervalScore = UnitIntervalScore(0.33);
+    const MIN_REQUIRED_EXTERIOR_PADDING_SCORE: UnitIntervalScore = UnitIntervalScore(0.5);
+
+    // interior marks could be cropped too, but less dramatically
+    const MIN_REQUIRED_INTERIOR_MARK_SCORE: UnitIntervalScore = UnitIntervalScore(0.7);
+    const MIN_REQUIRED_INTERIOR_PADDING_SCORE: UnitIntervalScore = UnitIntervalScore(0.7);
+
     complete_timing_marks
-        .bottom_rects
-        .par_iter()
```

Comment:
Previously this was doing a template match on each of the bottom timing marks, so I did it in parallel. Now we're just using the pre-computed scores so I removed the parallel iteration.